volatility3/framework/layers/qemu.py

```diff
@@ -69,6 +73,117 @@ def _read_configuration(self, base_layer: interfaces.layers.DataLayerInterface,
+        # | PC i440FX + PIIX "Old Default" | qemu/hw/i386/pc_piix.c:98      | 0xe0000000 | 0x100000000 |
+        # | PC Q35 + ICH9                  | qemu/hw/i386/pc_q35.c:141      | 0x80000000 | 0x100000000 |
+        # | MicroVM                        | qemu/hw/i386/microvm.c:291     | 0xc0000000 | 0x100000000 |
+        # | Xen                            | qemu/hw/i386/xen/xen-hvm.c:248 | 0xf0000000 | 0x100000000 |
+        # +--------------------------------+--------------------------------+------------+-------------+
+        #
+        # MicroVM and Xen haven't been implemented yet
+        #
+        # For now, we assume that the parameter max-ram-below-4g is not set, since this parameter influences the size
+        # and location of the memory gap. Deviating hole sizes could eventually be detected for Linux by e.g. scanning
+        # for dmesg entries with a regex like rb'\[mem (0x[0-9a-f]{4,10})-0x[0-9a-f]{4,10}\] available for PCI devices'
+
+        end = 0x100000000
+
+        if architecture:
+            if architecture[0] == 'i440fx':
+                if architecture[1] and architecture[1] < '2.0':
+                    start = 0xe0000000
+                else:
+                    start = 0xc0000000
+            elif architecture[0] == 'q35':
+                if ram_size >= 0xb0000000:
+                    start = 0x80000000
+                else:
+                    vollog.log(constants.LOGLEVEL_VV, f"No PCI-hole present in this memory layout "
+                                                      f"(Q35 + RAM size < 0xb0000000)")
+                    return None
+            else:
+                vollog.warning(f"Default location of PCI-hole not known for target architecture {architecture[0]}")
+                return None
+
+            vollog.log(constants.LOGLEVEL_VV, f"PCI-hole for detected target architecture is usually located at "
+                                              f"0x{start:x}, using this location")
+            return start, end - start
+        else:
+            vollog.warning("Could not determine target architecture, the possible presence of a PCI-hole is not taken "
+                           "into account")
+            return None
+
+    def _determine_architecture(self, base_layer: interfaces.layers.DataLayerInterface) -> \
+            Optional[Tuple[str, Optional[str]]]:
+        """Tries to determine the QEMU target architecture being used
+
+        Args:
+            base_layer: Layer to operate on
+
+        Returns:
+            (Optionally) A tuple of strings containing the name (in lowercase) and (optionally) the version of the used
+                QEMU target architecture
+        """
+        # MicroVM and Xen haven't been implemented yet
+
+        # First try to find a string containing both architecture and version
+        architecture_pattern = rb'pc-(i440fx|q35)-([0-9]{1,2}.[0-9]{1,2}(?:.[0-9]{1,2})?)'
```

Comment:
This information *has* to be in the qevm file (probably in the configuration), otherwise how would the memory be restored properly by qemu?